pkg/services/ngalert/schedule/schedule.go

```diff
@@ -486,21 +485,24 @@ func (sch *schedule) ruleRoutine(grafanaCtx context.Context, key ngmodels.AlertR
 				}()
 
 				err := retryIfError(func(attempt int64) error {
-					newVersion := ctx.rule.Version
 					isPaused := ctx.rule.IsPaused
-					// fetch latest alert rule version
-					if currentRuleVersion != newVersion {
-						// Do not clean up state if the eval loop has just started.
-						// We need to reset state if the loop has started and the alert is already paused. It can happen,
-						// if we have an alert with state and we do file provision with stateful Grafana, that state
-						// lingers in DB and won't be cleaned up until next alert rule update.
-						if currentRuleVersion > 0 || isPaused {
-							logger.Debug("Got a new version of alert rule. Clear up the state and refresh extra labels", "version", currentRuleVersion, "newVersion", newVersion)
-							resetState(grafanaCtx, isPaused)
-						}
-						currentRuleVersion = newVersion
+					f := ruleWithFolder{ctx.rule, ctx.folderTitle}.Fingerprint()
+					// Do not clean up state if the eval loop has just started.
+					var needReset bool
+					if currentFingerprint != 0 && currentFingerprint != f {
+						logger.Debug("Got a new version of alert rule. Clear up the state", "fingerprint", f)
+						needReset = true
+					}
+					// We need to reset state if the loop has started and the alert is already paused. It can happen,
+					// if we have an alert with state and we do file provision with stateful Grafana, that state
+					// lingers in DB and won't be cleaned up until next alert rule update.
+					needReset = needReset || (currentFingerprint == 0 && isPaused)
```

Comment:
I did that intentionally, which I think is clearer: basically the logic is that:
- do not reset when the rule routine has just started (it's fingerprint, used to be version is 0), reset only if version (and now fingerprint are different).
- however, if it's just started and the first eval isPaused=true then reset the state because of provisioning that does not reset the state of rule when it is provisioned.